libs/wampcc/wamp_session.cc

```diff
@@ -877,47 +872,71 @@ void wamp_session::handle_HELLO(json_array& ja)
     throw auth_error(WAMP_ERROR_AUTHENTICATION_FAILED,
                      "no auth methods available");
 
-  /* For now the only supported method is 'wampcra'. */
-  if (intersect.find(WAMP_WAMPCRA) == intersect.end())
-    throw auth_error(WAMP_ERROR_AUTHENTICATION_FAILED,
-                     "wampcra not available for both client and server");
+  /* Selecte the first method in the set as the common one */
+  std::string authmethod;
+  for(auto &element : intersect) {
+    authmethod = element;
+    break;
+  }
 
-  /* --- Perform wampcra authentication --- */
+  LOG_INFO(m_log_prefix << "selected '" << authmethod << "', as the common authentication method");
 
-  json_object challenge;
-  challenge["nonce"] = random_ascii_string(30);
-  challenge["authprovider"] = m_auth_proivder.provider_name(realm);
-  challenge["authid"] = authid;
-  challenge["timestamp"] = iso8601_utc_timestamp();
-  challenge["authrole"] = "user";
-  challenge["authmethod"] = WAMP_WAMPCRA;
-  challenge["session"] = std::to_string( unique_id() );
-  std::string challengestr = json_encode( challenge );
+  json_object extra;
 
-  {
-    std::lock_guard<std::mutex> guard(m_realm_lock);
-    if (m_challenge.empty())
-      m_challenge = challengestr;
-    else
-      throw auth_error(WAMP_ERROR_AUTHENTICATION_FAILED,
-                       "challenge already issued");
-  }
+  if(m_auth_proivder.hello) {
+    extra = m_auth_proivder.hello(authid, realm, authmethod, m_auth_proivder.provider_name(realm), unique_id() );
```

Comment:
Hi, what's the use case for the auth_provider to be able to provide own implementation of AUTHENTICATE.Extra  ?  E.g., if auth method is wampcra, then wouldn't they always have to use code like that contained in the lines 892 to 900 ?

Reply:
In case of `wampcra` it is not needed but for both [`cryptosign`](https://github.com/crossbario/crossbar/blob/master/crossbar/router/auth/cryptosign.py#L188) and [`scram`](https://github.com/crossbario/crossbar/blob/master/crossbar/router/auth/scram.py#L116) this would be needed.

With the ability to provide `extra` parameter for hello message in `auth_provider` it is possible to implement the missing authmethods without the need to change `wampcc` source code directly. 

With `hello` one can even re-implement `wampcra` for what ever reason.

Reply:
Okay ... just wondering whether this commit should onyl have the changes relevant to 'ticket', and a later commit can have support for custom auth handling etc.

Reply:
Another thought; for support for custom authentication feature, it might be necessary for classes like auth_error, to be moved to public access, so that they can be thrown during handling.

Reply:
It would be good to have it public. Do you want me to move it in this branch?

Reply:
Um... let me have another look through ... there might be other things to consider for having have custom hello/authenticate etc.

Reply:
Hi ... sorry for long delay. Will merge.  There are some changes I will make, but easier to do post merge. Thanks for commit.